File: Sheets/src/neetcode/practice/Backtracking/_1_subsets.java
package neetcode.practice.Backtracking;

import java.util.ArrayList;
import java.util.List;

class Solution {
  final public List<List<Integer>> result = new ArrayList<>();
  final public List<Integer> subset = new ArrayList<>();

  public List<List<Integer>> subsets(int[] nums) {
    this.dfs(nums, 0);
    return this.result;
  }

  private void dfs(int[] nums, int index) {
    if (index == nums.length) {
      this.result.add(new ArrayList<>(this.subset));
      return;
    }

    this.subset.add(nums[index]);
    this.dfs(nums, index + 1);
    this.subset.remove(this.subset.size() - 1);
    this.dfs(nums, index + 1);

  }
}

File: Sheets/src/neetcode/practice/Backtracking/_1_Subsets.cc
#include <vector>

class Solution {
public:
  std::vector<std::vector<int>> powerSet{};
  std::vector<int> subset{};

  void search(unsigned int idx, std::vector<int> &nums, std::size_t n) {
    if (idx == n)
      this->powerSet.push_back(this->subset);
    else {
      search(idx + 1, nums, n);
      this->subset.push_back(nums[idx]);
      search(idx + 1, nums, n);
      this->subset.pop_back();
    }
  }

  std::vector<std::vector<int>> subsets(std::vector<int> &nums) {
    search(0, nums, nums.size());
    return this->powerSet;
  }
};
